src/storage_versioned/transaction_versioned.rs

```diff
@@ -0,0 +1,70 @@
+use crate::TransactionInternal;
+use rocksdb::{TransactionDB, Error, ColumnFamily};
+use crate::common::{InternalReader, Reader, InternalRef};
+use crate::common::transaction::TransactionBasic;
+use crate::storage_versioned::StorageVersioned;
+use rocksdb::transactions::ops::GetColumnFamilies;
+
+pub struct TransactionVersioned<'a> {
+    transaction: TransactionInternal,           // TransactionInternal started for DB in CurrentState or for some version (checkpoint) of a StorageVersioned
+    storage_opt: Option<&'a StorageVersioned>,  // needed to create new version of DB in CurrentState after successful commit
+    db_opt:      Option<TransactionDB>          // needed to keep opened a DB for a specific version, otherwise the DB will be dropped (i.e. closed) after the end of the constructor's scope
+}
+
+impl InternalRef for TransactionVersioned<'_> {
+    fn db_ref(&self) -> Option<&TransactionDB> { None }
+    fn db_ref_mut(&mut self) -> Option<&mut TransactionDB> { None }
+
+    fn transaction_ref(&self) -> Option<&TransactionInternal> { Some(&self.transaction) }
+    fn transaction_ref_mut(&mut self) -> Option<&mut TransactionInternal> { Some(&mut self.transaction)  }
+}
+
+impl InternalReader for TransactionVersioned<'_> {}
+impl Reader for TransactionVersioned<'_> {}
+impl TransactionBasic for TransactionVersioned<'_> {}
+
+impl<'a> TransactionVersioned<'a> {
+    // Creates new instance of TransactionVersioned (which is a wrapper for TransactionInternal) for:
+    // - CurrentState of StorageVersioned if 'storage_opt is 'Some' and 'db_opt' is 'None' ('storage_opt' contains a reference to an instance of StorageVersioned for which the TransactionInternal is created);
+    // - some previous version of a StorageVersioned if 'storage_opt' is 'None' and 'db_opt' is 'Some' ('db_opt' contains an opened TransactionDB for some version of a StorageVersioned and a TransactionInternal is created for this DB);
+    // NOTE: storage_opt and db_opt can't be both specified as well as being both not specified
+    // Returns Result with a new instance of TransactionVersioned or with an error message if some error occurred
+    pub(crate) fn new(transaction: TransactionInternal, storage_opt: Option<&'a StorageVersioned>, db_opt: Option<TransactionDB>) -> Result<Self, Error> {
```

Comment:
We may use `Either` instead of to `Option` fields. That will prevent us to do the checks bellow.

Reply:
Fixed